designs/kindless-upgrades.md

```diff
@@ -0,0 +1,150 @@
+            controller->>infra: Check upgrade progress
+            controller->>+apiserver: Update cluster status
+            apiserver->>-controller: OK
+        end
+    and
+        loop Check for upgrade complete
+            CLI->>+apiserver: Get cluster status
+            apiserver->>-CLI: Cluster status
+        end
+    end
+    deactivate controller
+    
+    CLI->>-User: Cluster upgraded
+```
+
+### Validation webhook
+
+The current validation webhook prevents management clusters from being upgraded directly through the API. We need to change this to allow the CLI to leverage the controller for the upgrade.
+
+This presents the question: should we allow users to leverage the API to upgrade management clusters or should this be restricted to the CLI? This conditional limitation can be implemented using an annotation (for example the existing `managed-by-cli`). It's left up to the implementer to document pros and cons and help the team reach an agreement.
+
+### Cluster status
+
+In order for the CLI to track the progress made by the controller, we will use the `Cluster` object status. This will allow the CLI to not only show the current state of the upgrade process to the user but more importantly, to block until the upgrade process is completed.
+
+This will have its own design doc, which is already in progress.
+
+### Upgrade workflow changes
+
+The upgrade workflow needs to be updated to reflect this new architecture:
+
+* All bootstrap cluster operations and adjacent code should be removed.
+* The "upgrade needed" step is not needed and can be removed.
+* Pausing the EKS-A cluster needs to be removed, however, pausing the Flux controller is still needed.
+* Cilium upgrade needs to be removed from the core components upgrade task, since the controller takes care of this.
+* The upgrade task itself needs to be re-implemented. We recommend implementing in outside of the workflow, using the [new workflow tasks](https://github.com/aws/eks-anywhere/blob/main/pkg/workflow/task/workload/create.go) as inspiration. It should follow the following steps:
+
+    1. Update EKS-A resources with the user provided user spec using a Kubernetes client.
+    2. Watch/pull the cluster status, following the upgrade process. Beware of race conditions/caches and make sure we leverage `generation`/`observedGeneration` appropriately.
+    3. Blocking until the cluster is `Ready` or timeout is reached.
+
+* `reconcileClusterDefinitions` needs to be split and some parts reordered.
+    * Creating the new `Bundles` and EKS-D manifest in the cluster (in case of EKS-A version upgrade) needs to be moved up in the workflow, before core components upgrade.
+    * Updating the EKS-A resources is not needed, since the upgrade task already does it.
+    * All the GitOps operations should remain in this task.
+* Consider keeping a task to perform a backup of CAPI resources before applying any changes to the EKS-A cluster spec.
```

Comment:
Also during the upgrade flow in between the different phases?

Reply:
That's a good idea, however, that would be a new feature, since the CLI doesn't do that today. That said, it's not trivial to implement, since the phases are controlled by the controller. The CLI would somehow need to hook into those.

My 2c here would be if we want to go after this, discuss the feature separately (even in parallel) from the effort to remove the kind cluster.